feat(OptionsMenu): adds options menu

mcoker opened the pull request:
fixes https://github.com/patternfly/patternfly-next/issues/1088

Also added a small fix for the expanded color on the app launcher @matthewcarleton

patternfly-build commented:
Deploy preview for *pf-next* ready!

Built with commit cc5cdec883ab8798321fd971d5c20ace11798082

https://deploy-preview-1264--pf-next.netlify.com

mcoker commented:
@matthewcarleton nope, I don't think the app launcher needs those, either.

jgiardino commented on src/patternfly/components/OptionsMenu/examples/options-menu-single-example.hbs:
```diff
@@ -0,0 +1,19 @@
+{{#> options-menu id="options-menu-single-example" options-menu--label-text="Choose number of items per page"}}
```
Can we change this to something like "Page size" or "Items per page"? I like to think of these labels the same way we think about labels in a form, and we would use a label like "Choose number of..." for a select or "Enter your..." for a text input. The semantics provided by the aria attributes on the toggle communicate to screen reader users that they will be presented with choices when they act on the toggle.

jgiardino commented on src/patternfly/components/OptionsMenu/examples/options-menu-align-right-example.hbs:
```diff
@@ -0,0 +1,6 @@
+{{#> options-menu options-menu--IsExpanded="true" id="options-menu-multiple-example" options-menu--label-text="Choose sort methods"  options-menu--modifier="pf-m-align-right"}}
```
Similar to previous comment, we don't need the word "Choose" as part of this label. Also, this raises an interesting question of how do differentiate the sub-menus and associate labels with the sub-menus. I can try to find examples of this that we could follow, but I'm wondering if those labels would be "Sort direction" and "Sort by"?

jgiardino replied:
The Style/Color in this [wai-aria menu example](https://www.w3.org/TR/wai-aria-practices-1.1/examples/menubar/menubar-2/menubar-2.html) is similar to what we have. They use nested lists, where the primary list is labeled by the toggle, and the nested lists have labels that describe their group of options (e.g. "Sort direction" and "Sort by" in our case).

matthewcarleton commented:
🤦‍♂️ looking at this on my phone I for some reason thought it was adding it to app launcher. Ignore my previous comment.

christiemolloy commented on src/patternfly/components/OptionsMenu/options-menu.scss:
```diff
@@ -0,0 +1,201 @@
+  --pf-c-options-menu--m-top__menu--Transform: translateY(calc(-100% - var(--pf-global--spacer--xs))); // The "dropup" menu must be transformed up and this calculates how much to create space between the toggle and menu
+
+  // Menu Item
+  --pf-c-options-menu__menu-item--PaddingTop: var(--pf-global--spacer--sm);
```
The spacing looks too small between the spacer and list item above and below it. Did you follow these spacing guidelines? 
<img width="354" alt="screen shot 2019-01-21 at 11 00 21 am" src="https://user-images.githubusercontent.com/20118816/51485386-e740ca80-1d6b-11e9-851d-896643b9c58d.png">

mcoker replied:
Oh I'm just using the dropdown's spacing. I'm assuming this should match the dropdown? Or should the dropdown (and select) menu item spacing be changed you think?

christiemolloy replied:
@gdoyle1 can you double check this

gdoyle1 replied:
@mcoker @christiemolloy It should match the dropdown spacing but I think the space between the spacer and list item needs to be increased in general. Could we do that?

mcoker replied:
@gdoyle1 yep, just let me know what you'd like it to be.

gdoyle1 replied:
@mcoker 16 px below "size" then the divider line then 16 px above "ascending" please!

mcoker replied:
thanks @gdoyle1 :) Is this just for this type of menu? or is this for all dropdowns with a separator? I copied these styles from the dropdown - https://pf-next.com/components/Dropdown/examples/

gdoyle1 replied:
@mcoker For all dropdowns with a separator. Thank you!

mcoker replied:
Awesome, thanks. Updated options-menu, dropdown, and select.

gdoyle1 replied:
@mcoker That spacing looks a lot better! Thank you

christiemolloy commented on src/patternfly/components/OptionsMenu/options-menu.scss:
```diff
@@ -0,0 +1,201 @@
+  &.pf-m-hover,
+  &:focus,
+  &.pf-m-focus {
+    --pf-c-options-menu__menu-item--Color: var(--pf-c-options-menu__menu-item--Color);
```
should this just be `color: var(--pf-c-options-menu__menu-item--Color)` ? I don't see `--pf-c-options-menu__menu-item--Color` being defined before?

mcoker replied:
good catch, looks like this isn't needed, and the disabled color wasn't being applied either.

mcoker commented on src/patternfly/components/OptionsMenu/docs/code.md:
```diff
@@ -0,0 +1,26 @@
+| Class | Applied To | Outcome |
+| -- | -- | -- |
+| `.pf-c-options-menu` | `<div>` |  Initiates a custom options menu. |
+| `.pf-c-options-menu__toggle` | `<button>` |  Initiates a custom toggle. |
```
It can be a `div` when the whole toggle shouldn't be interactive, and there is an interactive element (or elements) inside. I didn't think that was the case here, but per Jenn's comment above, I think this is going to support being a `div`, too

jgiardino replied:
Oooh, the select should be using a `<button>`. I captured that in https://github.com/patternfly/patternfly-next/issues/1280

mcoker commented:
* updated pagination example to have a separate button from the text
* updated labels to be descriptive, rather than stating an action
* nested multiple options example to have multiple lists, with the main list label on the button, and the second list as a nested `ul` with an `aria-label`

jgiardino commented:
Your updates look great! Just a few minor suggestions...

The first group of options uses first-level list items:
```
<ul class="pf-c-options-menu__menu" aria-labelledby="options-menu-multiple-example-label">
    <li>
      <button class="pf-c-options-menu__menu-item">
        Name
      </button>
    </li>
```

Whereas the second group of options uses second-level list items:
```
<ul class="pf-c-options-menu__menu" aria-labelledby="options-menu-multiple-example-label">
  ...
  <li>
      <ul aria-label="Sort direction">
        <li>
          <button class="pf-c-options-menu__menu-item pf-m-selected" aria-selected="true">
            Ascending
            <i class="fas fa-check pf-c-options-menu__menu-item-icon" aria-hidden="true"></i>
          </button>
        </li>
```
Since only one option can be selected in the first group, it should use the same structure as the second group (`<ul>` > `<li>` > `<ul aria-label="Sort by">` > `<li>`).

jgiardino commented:
I noticed that the hover and active styles aren't the same as those used for Dropdown and Select components. There is an issue with first couple of examples for Select (https://github.com/patternfly/patternfly-next/issues/1280), but there are some other examples, like **Select - Multi with Typeahead**, that are comparable to this component, where the wrapping element is a div, and when a child element receives focus, the indication is applied to the whole element.

mcoker commented:
Awesome, thanks so much for the reviews @jgiardino :D Made these updates.

srambach commented:
I notice in the first example, I don't see the hover until I'm over the caret. Will the user expect to be able to click the entire thing to open it?

mcoker commented:
@srambach per jenn's comment

> the cleanest solution at this point seems to be just separating the text "1-10 of 25 Items" from the select menu.

and that the hover indication should only apply when you're on the clickable thing, it seems to me that you wouldn't show the hover indication until you hover the button itself.

srambach commented:
@mcoker my apologies, I hadn't followed the whole the discussion above. @jgiardino so by "separating" you meant that the caret would be clickable but the text is not? Or do you mean to visually separate? I personally think it looks as if it should all be clickable, especially since hovering on the caret makes the entire underline highlight.

jgiardino commented:
> by "separating" you meant that the caret would be clickable but the text is not?

By separating, I mean that the caret would be a button that does not include the visible text.

> I personally think it looks as if it should all be clickable, especially since hovering on the caret makes the entire underline highlight.
Yes, I agree. The visual styling also conveys to me that the whole thing is clickable. 

I think this is more of a question for @mcarrano and someone from visual design (would that be @mceledonia?), and maybe @stacymcauliffe?

It might be possible to provide a single label for a select menu toggle button that includes:

- which items are listed out of total number of items
- page size options are available 

But I feel like the cleanest solution is to separate these pieces of information in the DOM. And if that's the case, then maybe they should be visually separate pieces as well.

mcarrano commented:
WRT the question about what should be clickable, I compared this against the React implementation of the Dropdown here: http://patternfly-react.surge.sh/patternfly-4/components/dropdown Hovering anywhere on the text or the caret displays the hover state and single click anywhere toggles the menu. This feels right to me. I expect that the Dropdown and the Options menu would have the same interaction other than the menu itself. Do others agree?

gdoyle1 commented:
@mcarrano I agree that the dropdown and options menu would have the same interaction on hover and select (other than the menu itself). So agreeing with "hovering the text or caret would display the hover state and a single click anywhere toggles the menu"

mcoker commented:
@mcarrano @srambach @gdoyle1 @jgiardino I added another example that shows the actual pagination options menu. I've left that variation with the text as separate from the button, so only the button is clickable, but there is no border so it's less confusing. I've updated the other examples so that the entire toggle area is clickable when indicated with the standard dropdown border treatment.

mcarrano commented:
This looks good @mcoker . I agree that the options menu in pagination is somewhat unique from the others. 

The only thing I'm wondering is if the examples are misleading. It seems to imply that a single select drops from a text-based button and the multi from an icon. In fact the menu interaction (single or multi) should be independent from what is used for the toggle. So the Sort menu, for example, could just as well be dropped from a button labeled Sort as it could from the icon. I'm fine as long as we are not constraining this and you think that the above would be clear to consumers of the component.

christiemolloy commented on src/patternfly/components/OptionsMenu/options-menu.scss:
```diff
@@ -0,0 +1,225 @@
+    right: 0;
+  }
+
+  .pf-c-options-menu.pf-m-top & {
```
for pf-m-top modifier, should the arrow be facing upwards towards the menu, I see this in the dropdown component

christiemolloy commented:
@mcoker left a comment

mcoker commented:
@mcarrano agreed, I combined the plain example into the multiple example to save time. Separated those out so we have a normal multiple options menu example and a plain example.

patternfly-build commented:
:tada: This PR is included in version 1.0.148 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/@patternfly/patternfly-next)
- [GitHub release](https://github.com/patternfly/patternfly-next/releases/tag/v1.0.148)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
